Compiler suggested <unnamed> in a use

```
error[E0599]: no method named `size` found for type `glutin_window::GlutinWindow` in the current sco
pe
  --> src/grenderer/game.rs:28:39
   |
28 |             let tilespace_y = (window.size().height / 4) * 3;
   |                                       ^^^^
   |
   = help: items from traits can only be used if the trait is in scope
   = note: the following trait is implemented but not in scope, perhaps add a `use` for it:
           candidate #1: `use piston::<unnamed>::Window;`

```
Attempting to use that suggustion causes a compiler error itself, and the correct use statement is `use piston::window::Window;`

I dont know how to reproduce this behavior, but it can likely be found by looking in the piston_window crate's source code.

Looks like it's because `piston::window` is an `extern crate`. Simplified:

```sh
#!/bin/sh
set -ex
rm -rf a b c
cargo new --lib a
cargo new --lib b
cargo new --bin c
echo 'pub trait A { fn a(&self) {} } impl A for () {}' > a/src/lib.rs
echo 'a = { path = "../a" }' >> b/Cargo.toml
echo 'pub extern crate a;' > b/src/lib.rs
echo 'b = { path = "../b" }' >> c/Cargo.toml
echo 'extern crate b; fn main() { ().a(); }' > c/src/main.rs
cargo build --manifest-path c/Cargo.toml
```

Running this gives the error:

```
+ rm -rf a b c
+ cargo new --lib a
     Created library `a` project
+ cargo new --lib b
     Created library `b` project
+ cargo new --bin c
     Created binary (application) `c` project
+ echo 'pub trait A { fn a(&self) {} } impl A for () {}'
+ echo 'a = { path = "../a" }'
+ echo 'pub extern crate a;'
+ echo 'b = { path = "../b" }'
+ echo 'extern crate b; fn main() { ().a(); }'
+ cargo build --manifest-path c/Cargo.toml
   Compiling a v0.1.0 (file:///Users/kennytm/Downloads/a/a)
   Compiling b v0.1.0 (file:///Users/kennytm/Downloads/a/b)
   Compiling c v0.1.0 (file:///Users/kennytm/Downloads/a/c)
error[E0599]: no method named `a` found for type `()` in the current scope
 --> c/src/main.rs:1:32
  |
1 | extern crate b; fn main() { ().a(); }
  |                                ^
  |
  = help: items from traits can only be used if the trait is in scope
  = note: the following trait is implemented but not in scope, perhaps add a `use` for it:
          candidate #1: `use b::<unnamed>::A;`

error: aborting due to previous error

error: Could not compile `c`.

To learn more, run the command again with --verbose.
```

(Here crate `a` is `pistoncore-window`, crate `b` is `piston`, and crate `c` is the user's crate.)

--- a/src/librustc/ty/item_path.rs
+++ b/src/librustc/ty/item_path.rs
@@ -169,7 +169,11 @@ impl<'a, 'gcx, 'tcx> TyCtxt<'a, 'gcx, 'tcx> {
 
             let data = cur_def_key.disambiguated_data.data;
             let symbol = data.get_opt_name().map(|n| n.as_str()).unwrap_or_else(|| {
-                Symbol::intern("<unnamed>").as_str()
+                if let DefPathData::CrateRoot = data {  // reexported `extern crate` (#43189)
+                    self.original_crate_name(cur_def.krate).as_str()
+                } else {
+                    Symbol::intern("<unnamed>").as_str()
+                }
             });
             cur_path.push(symbol);
 

--- a/src/test/ui/auxiliary/xcrate_issue_43189_a.rs
+++ b/src/test/ui/auxiliary/xcrate_issue_43189_a.rs
@@ -1,0 +1,17 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#![crate_type="lib"]
+
+
+pub trait A {
+    fn a(&self) {}
+}
+impl A for () {}

--- a/src/test/ui/auxiliary/xcrate_issue_43189_b.rs
+++ b/src/test/ui/auxiliary/xcrate_issue_43189_b.rs
@@ -1,0 +1,13 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+#![crate_type="lib"]
+
+pub extern crate xcrate_issue_43189_a;

--- a/src/test/ui/issue-43189.rs
+++ b/src/test/ui/issue-43189.rs
@@ -1,0 +1,22 @@
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// Issue 46112: An extern crate pub re-exporting libcore was causing
+// paths rooted from `std` to be misrendered in the diagnostic output.
+
+// ignore-windows
+// aux-build:xcrate_issue_43189_a.rs
+// aux-build:xcrate_issue_43189_b.rs
+
+extern crate xcrate_issue_43189_b;
+fn main() {
+    ().a();
+    //~^ ERROR no method named `a` found for type `()` in the current scope [E0599]
+}

--- a/src/test/ui/issue-43189.stderr
+++ b/src/test/ui/issue-43189.stderr
@@ -1,0 +1,15 @@
+error[E0599]: no method named `a` found for type `()` in the current scope
+  --> $DIR/issue-43189.rs:20:8
+   |
+LL |     ().a();
+   |        ^
+   |
+   = help: items from traits can only be used if the trait is in scope
+help: the following trait is implemented but not in scope, perhaps add a `use` for it:
+   |
+LL | use xcrate_issue_43189_b::xcrate_issue_43189_a::A;
+   |
+
+error: aborting due to previous error
+
+For more information about this error, try `rustc --explain E0599`.

